scopa.py

```diff
@@ -0,0 +1,217 @@
+
+        # iterate through possible takes and check what is the best - if any
+        best_take_score = -1
+        best_take_index = no_take
+        for t in range(0, len(potential_takes)):
+            # remember the last hand played
+            self.last_hand_played = hand_number
+            # check what is the score
+            current_take_score = tactics.get_score_for_take(potential_takes[t], self.table)
+            if current_take_score > best_take_score:
+                best_take_index = t
+                best_take_score = current_take_score
+
+        # best_take_index will contain the index of best take
+        # if there is any take, take the best
+        if best_take_index != no_take:
+            # remove card from hand and add it to the pile
+            card_from_hand = potential_takes[best_take_index][0]
+            self.hands[hand_number].remove(card_from_hand)
+            self.piles[hand_number].append(card_from_hand)
+            # remove cards from the table and add them to the pile
+            for card in potential_takes[best_take_index][1]:
+                self.table.remove(card)
+                self.piles[hand_number].append(card)
+
+            # if no cards left on the table, increase scopa count
+            if len(self.table) == 0 and not len(self.deck) == 0:
+                self.scopa_count[hand_number] += 1
+
+            return potential_takes[best_take_index]
+
+        else:
+            # no cards can be taken, drop a card
+            card_to_be_dropped = tactics.select_card_to_be_dropped(self.hands[hand_number], self.table)
+            self.hands[hand_number].remove(card_to_be_dropped)
+            self.table.append(card_to_be_dropped)
+            return [card_to_be_dropped, []]
+
+    # this function calculates results of a pile with given number
+    def calculate_score(self, pile_number):
+        score = 0
+        # one point for settebello
+        if tactics.settebello(self.piles[pile_number]):
+            score += 1
+
+        # check if any other hand has more or the same number of sevens, cards and denars
+        no_of_sevens = tactics.sevens(self.piles[pile_number])
+        more_sevens = True
+        no_of_cards = len(self.piles[pile_number])
+        more_cards = True
+        no_of_denars = tactics.denars(self.piles[pile_number])
+        more_denars = True
+        for i in range(0, len(self.piles)):
+            if i != pile_number:
+                if tactics.sevens(self.piles[i]) >= no_of_sevens:
+                    more_sevens = False
+                if len(self.piles[i]) >= no_of_cards:
+                    more_cards = False
+                if tactics.denars(self.piles[i]) >= no_of_denars:
+                    more_denars = False
+
+        # add one point for each
+        if more_sevens:
+            score += 1
+        if more_cards:
+            score += 1
+        if more_denars:
+            score += 1
```

Comment:
Since the loops are short (<=40 iterations) we can separate checking for points in the pile. Extracting and accordingly naming the methods would be more readable.